File: autolink/mmkv/android/src/main/java/com/lynxapp/autolink/mmkv/NativeKVModule.java
package com.lynxapp.autolink.mmkv;

import androidx.annotation.Nullable;

import com.lynx.react.bridge.Callback;
import com.lynx.jsbridge.LynxContextModule;
import com.lynx.jsbridge.LynxMethod;
import com.lynx.jsbridge.LynxNativeModule;
import com.lynx.tasm.behavior.LynxContext;
import com.tencent.mmkv.MMKV;

import java.util.concurrent.atomic.AtomicBoolean;

/** String primitives shared by every bundle; JSON encoding stays in TypeScript. */
@LynxNativeModule(name = NativeKVModule.NAME)
public class NativeKVModule extends LynxContextModule {
    public static final String NAME = "NativeKVModule";
    private static final String STORAGE_ID = "lynx.native.kv";
    private static final AtomicBoolean INITIALIZED = new AtomicBoolean(false);

    private final MMKV storage;

    public NativeKVModule(LynxContext context) {
        super(context);
        // The library owns its MMKV bootstrap so hosts no longer call
        // MMKV.initialize() themselves; the call is idempotent per process.
        if (INITIALIZED.compareAndSet(false, true)) {
            MMKV.initialize(context);
        }
        storage = MMKV.mmkvWithID(STORAGE_ID);
    }

    @LynxMethod
    public void setString(String key, String value, Callback callback) {
        boolean ok = isValidKey(key) && storage.encode(key, value);
        callback.invoke(ok ? "" : "Unable to persist MMKV key");
    }

    @LynxMethod
    public void getString(String key, @Nullable String defaultValue, Callback callback) {
        if (!isValidKey(key)) {
            callback.invoke(defaultValue);
            return;
        }
        callback.invoke(storage.decodeString(key, defaultValue));
    }

    @LynxMethod
    public void remove(String key, Callback callback) {
        if (!isValidKey(key)) {
            callback.invoke("MMKV key must not be empty");
            return;
        }
        storage.removeValueForKey(key);
        callback.invoke("");
    }

    @LynxMethod
    public void clear(Callback callback) {
        storage.clearAll();
        callback.invoke("");
    }

    @LynxMethod
    public void contains(String key, Callback callback) {
        callback.invoke(isValidKey(key) && storage.containsKey(key));
    }

    private static boolean isValidKey(String key) {
        return key != null && !key.trim().isEmpty();
    }
}

File: app/androidApp/app/src/main/java/com/lynxapp/nativemodule/NativeKVModule.kt
package com.lynxapp.nativemodule

import android.content.Context
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.tencent.mmkv.MMKV

/** String primitives shared by every bundle; JSON encoding stays in TypeScript. */
class NativeKVModule(context: Context) : LynxModule(context) {
    private val storage: MMKV = requireNotNull(MMKV.mmkvWithID(STORAGE_ID)) {
        "Unable to open MMKV storage $STORAGE_ID"
    }

    @LynxMethod
    fun setString(key: String, value: String, callback: Callback) {
        val error = if (isValidKey(key) && storage.encode(key, value)) {
            ""
        } else {
            "Unable to persist MMKV key"
        }
        callback.invoke(error)
    }

    @LynxMethod
    fun getString(key: String, defaultValue: String?, callback: Callback) {
        if (!isValidKey(key)) {
            callback.invoke(defaultValue)
            return
        }
        callback.invoke(storage.decodeString(key, defaultValue))
    }

    @LynxMethod
    fun remove(key: String, callback: Callback) {
        if (!isValidKey(key)) {
            callback.invoke("MMKV key must not be empty")
            return
        }
        storage.removeValueForKey(key)
        callback.invoke("")
    }

    @LynxMethod
    fun clear(callback: Callback) {
        storage.clearAll()
        callback.invoke("")
    }

    @LynxMethod
    fun contains(key: String, callback: Callback) {
        callback.invoke(isValidKey(key) && storage.containsKey(key))
    }

    private fun isValidKey(key: String): Boolean = key.isNotBlank()

    companion object {
        const val NAME = "NativeKVModule"
        private const val STORAGE_ID = "lynx.native.kv"
    }
}
